### ISSUE TITLE
Pulse caching

### ISSUE THREAD

**Issue by caryan** [BBN-Q/QGL#54]
Now that we translate to `Waveforms` early on we should be able to reactivate the [memoization](https://github.com/BBN-Q/QGL/blob/2de8c17974a06ea9a0252936aca5c54295a82078/QGL/PulsePrimitives.py#L34) we've had commented out for 4! years.  We could also make Pulses explicitly immutable perhaps using a [namedtuple](https://docs.python.org/3/library/collections.html#collections.namedtuple).  The only issue I see is we need some hook to invalidate the cache when pulse parameters are changed by the file watcher.

**Comment by caryan** [comment]
As a reference point at  3970877 using the GST data file from #4 and a slightly modified script

```python
from QGL import *
import csv

q1 = QubitFactory("q1")
q2 = QubitFactory("q2")

gates = [(X90,q1), (X90,q2), (Y90,q1), (Y90,q2)]


with open("/home/cryan/Downloads/gst16-2q.csv", 'r') as csvfile:
	lines = csvfile.read().splitlines()

def create_seqs(lines):
	seqs = []
	for line in lines:
		seq = []
		for g in map(int, line.split(",")):
			if g == 5:
				seq += ZX90_CR(q1, q2)
			else:
				if g != 0:
					seq.append( gates[g-1][0](gates[g-1][1]) )

		seq.append( MEAS(q1)*MEAS(q2) )

		seqs.append(seq)
	return seqs

# filenames = compile_to_hardware(seqs, "Test/Test")

```

Using iPython's profiling e.g.

```python
In [2]: %timeit seqs = create_seqs(lines[:2000])
10 loops, best of 3: 179 ms per loop
```

or

```python
%%timeit
seqs = create_seqs(lines[:500])
compile_to_hardware(seqs, "Test/Test")
```

or

```
In [1]: %paste
from QGL import *
import csv

q1 = QubitFactory("q1")
q2 = QubitFactory("q2")

gates = [(X90,q1), (X90,q2), (Y90,q1), (Y90,q2)]


with open("/home/cryan/Downloads/gst16-2q.csv", 'r') as csvfile:
        lines = csvfile.read().splitlines()

## -- End pasted text --

In [2]: from silly import create_seqs

In [3]: %load_ext memory_profiler

In [4]: %mprun -f create_seqs seqs = create_seqs(lines[:2000])
Filename: /home/cryan/Programming/Repos/QGL/silly.py

Line #    Mem usage    Increment   Line Contents
================================================
    37    131.8 MiB      0.0 MiB   def create_seqs(lines):
    38    131.8 MiB      0.0 MiB   	seqs = []
    39    160.6 MiB     28.8 MiB   	for line in lines:
    40    160.6 MiB      0.0 MiB   		seq = []
    41    160.6 MiB      0.0 MiB   		for g in map(int, line.split(",")):
    42    160.6 MiB      0.0 MiB   			if g == 5:
    43    159.6 MiB     -1.1 MiB   				seq += ZX90_CR(q1, q2)
    44                             			else:
    45    160.6 MiB      1.1 MiB   				if g != 0:
    46    160.6 MiB      0.0 MiB   					seq.append( gates[g-1][0](gates[g-1][1]) )
    47
    48    160.6 MiB      0.0 MiB   		seq.append( MEAS(q1)*MEAS(q2) )
    49
    50    160.6 MiB      0.0 MiB   		seqs.append(seq)
    51    160.6 MiB      0.0 MiB   	return seqs


```

I get timings and memory on my laptop as follows.  Memory profiling on Pyhton
seems to be woefully inadequate so use the iPython process memory as a proxy.
Timings are skewed by #57  but taking "best of" should help

| Num. Seqs. | Time to Create | Memory Allocated  | Compile time (code saving) | iPython Memory |  Compile time (no saving)
| :---:      | :--:           |     :---:         |       :--:                 |     :---:      |       :---:
| 100        |  6ms           |    0B???          |     348ms                  |      94 MB     |       310ms
| 1000       |  74ms          |    11 MB          |      9.5s                  |      184 MB    |       8.4s
| 2000       |  179ms         |    221 MB         |      55s                   |      288 MB    |       30s
| 27017      |  4.9s          |    912 MB         |     ???                    |        ???     |       ????

**Comment by caryan** [comment]
Some significant improvements as of 5c2246f

```python
In [2]: %timeit create_seqs(lines)
1 loop, best of 3: 2.33 s per loop
```

```python
In [2]: %%timeit
seqs = create_seqs(lines[:2000])
compile_to_hardware(seqs, "Test/Test")
   ...: 
Compiled 2000 sequences.
Compiled 2000 sequences.
Compiled 2000 sequences.
Compiled 2000 sequences.
1 loop, best of 3: 5.66 s per loop
```

**Comment by caryan** [comment]
Closed by #60
